common/common.go

```diff
@@ -1,46 +1,200 @@
+	params[1] = json.RawMessage("0")
+	result, rpcErr := rpcClient.RawRequest("getblock", params)
 
 	// For some reason, the error responses are not JSON
 	if rpcErr != nil {
 		errParts := strings.SplitN(rpcErr.Error(), ":", 2)
-		errCode, err = strconv.ParseInt(errParts[0], 10, 32)
+		errCode, err := strconv.ParseInt(errParts[0], 10, 32)
+		// Check to see if we are requesting a height the zcashd doesn't have yet
 		if err == nil && errCode == -8 {
-			return -1, -1, "", "", nil
+			return nil, nil
 		}
-		return -1, -1, "", "", errors.Wrap(rpcErr, "error requesting block")
+		return nil, errors.Wrap(rpcErr, "error requesting block")
 	}
 
-	var f interface{}
-	err = json.Unmarshal(result, &f)
+	var blockDataHex string
+	err := json.Unmarshal(result, &blockDataHex)
 	if err != nil {
-		return -1, -1, "", "", errors.Wrap(err, "error reading JSON response")
+		return nil, errors.Wrap(err, "error reading JSON response")
 	}
 
-	chainName := f.(map[string]interface{})["chain"].(string)
+	blockData, err := hex.DecodeString(blockDataHex)
+	if err != nil {
+		return nil, errors.Wrap(err, "error decoding getblock output")
+	}
 
-	upgradeJSON := f.(map[string]interface{})["upgrades"]
-	saplingJSON := upgradeJSON.(map[string]interface{})["76b809bb"] // Sapling ID
-	saplingHeight := saplingJSON.(map[string]interface{})["activationheight"].(float64)
+	block := parser.NewBlock()
+	rest, err := block.ParseFromSlice(blockData)
+	if err != nil {
+		return nil, errors.Wrap(err, "error parsing block")
+	}
+	if len(rest) != 0 {
+		return nil, errors.New("received overlong message")
+	}
 
-	blockHeight := f.(map[string]interface{})["headers"].(float64)
+	return block.ToCompact(), nil
+}
 
-	consensus := f.(map[string]interface{})["consensus"]
-	branchID := consensus.(map[string]interface{})["nextblock"].(string)
+func BlockIngestor(rpcClient *rpcclient.Client, cache *BlockCache, log *logrus.Entry, startHeight int) {
+	reorgCount := 0
+	height := startHeight
+
+	// Start listening for new blocks
+	retryCount := 0
+	for {
+		block, err := getBlockFromRPC(rpcClient, height)
+		if block == nil || err != nil {
+			if err != nil {
+				log.WithFields(logrus.Fields{
+					"height": height,
+					"error":  err,
+				}).Warn("error with getblock rpc")
+				retryCount++
+				if retryCount > 10 {
+					log.WithFields(logrus.Fields{
+						"timeouts": retryCount,
+					}).Fatal("unable to issue RPC call to zcashd node")
+				}
+			}
+			// We're up to date in our polling; wait for a new block
+			time.Sleep(10 * time.Second)
+			continue
+		}
+		retryCount = 0
+
+		log.Info("Ingestor adding block to cache: ", height)
+		err, reorg := cache.Add(height, block)
+
+		if err != nil {
+			// It's unclear how this will recover, but we certainly
+			// don't want to loop full-speed
+			log.Error("Error adding block to cache: ", err)
+			time.Sleep(10 * time.Second)
+			continue
+		}
+
+		// Check for reorgs once we have inital block hash from startup
+		if reorg {
+			height -= 10
```

Comment:
This should probably be just -1 or -2. Re-reading 10 blocks for each re-org seems excessive.